### PR TITLE
L009: Handle adding newline after trailing templated code

### LINKED ISSUE #2046: dbt - L009: Files must end with a single trailing newline.
## Expected Behaviour
No error

## Observed Behaviour
```
L:  5 | P:  60 | L009 | Files must end with a single trailing newline.    
```

## Steps to Reproduce
A dbt model that calls a macro, eg:
```
{{ config ( materialized='incremental',
   transient=false,
    ) }}

{{ generate_table('mytable') }}

```

## Dialect
snowflake

## Version
sqlfluff, version 0.8.2
python 3.9

## Configuration
```
[tool.sqlfluff.core]
templater = "dbt"
sql_file_exts = ".sql,.sql.j2,.dml,.ddl"
dialect = "snowflake"
ignore_templated_areas = true

[tool.sqlfluff.indentation]
indented_joins = false
indented_using_on = true
template_blocks_indent = false

[tool.sqlfluff.templater.dbt]
project_dir = "."
profile = "snowflake"
```

### PR DESCRIPTION
**Pull request by barrywhart**
### Are there any other side effects of this change that we should be aware of?

Fixes #2046. (A previous PR that was thought to fix it was missing one case.)

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/linter/linted_file.py [hunk 1]
```diff
@@ -330,10 +330,11 @@ def fix_string(self) -> Tuple[Any, bool]:
                 source_str=self.templated_file.source_str[source_slice],
             )
 
-            # Deal with the easy case of only literals
-            if set(local_type_list) == {"literal"}:
+            # Deal with the easy cases of 1) New code at end 2) only literals
+            if not local_type_list or set(local_type_list) == {"literal"}:
```
Comment (barrywhart) [inline src/sqlfluff/core/linter/linted_file.py:334]: Prior to this fix, the linter was discarding the fix if we added a newline at the end of the file, immediately following a templated region.

File: src/sqlfluff/core/rules/functional/segment_predicates.py [hunk 2]
```diff
@@ -152,3 +153,30 @@ def raw_slices(
...
+    """Returns raw slices for a segment."""
+    if not templated_file:
+        raise ValueError(
+            'templated_slices: "templated_file" parameter is required.'
+        )  # pragma: no cover
+    templated_slices = []
+    # :TRICKY: We don't use _find_slice_indices_of_templated_pos() here because
+    # it treats TemplatedFileSlice.templated_slice.stop as inclusive, not
+    # exclusive. Other parts of SQLFluff rely on this behavior, but we don't
+    # want it. It's easy enough to do this ourselves.
```
Comment (jpy-git) [inline src/sqlfluff/core/rules/functional/segment_predicates.py:171]: Nice, I've encountered this issue before so good to get around it here 👍

File: test/fixtures/rules/std_rule_cases/L009.yml [hunk 3]
```diff
@@ -13,3 +13,7 @@ test_fail_multiple_final_newlines:
 
 test_pass_templated_plus_raw_newlines:
   pass_str: "{{ '\n\n' }}\n"
+
+test_fail_templated_plus_raw_newlines:
+  fail_str: "{{ '\n\n' }}"
+  fix_str: "{{ '\n\n' }}\n"
```
Comment (jpy-git) [inline test/fixtures/rules/std_rule_cases/L009.yml:19]: Could we also have a unit test for this please:
```yml
test_fail_templated_plus_raw_newlines_extra_newline:
  fail_str: "{{ '\n\n' }}\n\n"
  fix_str: "{{ '\n\n' }}\n"
```

Comment (jpy-git) [reply test/fixtures/rules/std_rule_cases/L009.yml:19]: think it works but good to confirm

**Review by jpy-git — APPROVED** [review]
Looks good, worth adding in that extra unit test for coverage, but otherwise happy 👍

**Pull request merged**